// minify/legacy/lib/html.js
/* сжимаем код через htmlMinify */

'use strict';

var assert = require('assert');
var tryCatch = require('try-catch');
var Minifier = require('html-minifier');

var ErrorMsg = Error(['can\'t load html-minifier', 'npm install html-minifier', 'https://github.com/kangax/html-minifier'].join('\n'));

var Options = {
    removeComments: true,
    removeCommentsFromCDATA: true,
    removeCDATASectionsFromCDATA: true,
    collapseWhitespace: true,
    collapseBooleanAttributes: true,
    removeAttributeQuotes: true,
    removeRedundantAttributes: true,
    useShortDoctype: true,
    removeEmptyAttributes: true,
    /* оставляем, поскольку у нас
     * в элемент fm генерируеться
     * таблица файлов
     */
    removeEmptyElements: false,
    removeOptionalTags: true,
    removeScriptTypeAttributes: true,
    removeStyleLinkTypeAttributes: true,

    minifyJS: true,
    minifyCSS: true
};

/**
 * minify html data.
 *
 * @param data
 * @param callback
 */
module.exports = function (data, callback) {
    assert(data);
    assert(callback);

    if (!Minifier) return callback(ErrorMsg);

    var error = tryCatch(function () {
        data = Minifier.minify(data, Options);
    });

    callback(error, data);
};
